File: server/Message/MessageModule.cs
//////////////////////////////////////////////////////////////////////////
// 
// 文件：server/Message/MessageModule.cs
// 作者：Xoen Xie
// 时间：2023/06/19
// 描述：消息协议模块
// 说明：
//
//////////////////////////////////////////////////////////////////////////

using XSF;
using Serilog;
using XsfPb;
using XsfPbid;

namespace XsfMsg
{
    public partial class MessageModule : IModule, IMessageHelper
    {
        private IMessage[] m_MessagePool;

        public MessageModule()
        {
            m_MessagePool = new IMessage[10000];

            // 下面的注释定义必须保留，用来自动生成代码
            //MESSAGE_START
			m_MessagePool[(int)CMSGID.CltGtHandshake] = new MSG_Clt_Gt_Handshake();
			m_MessagePool[(int)CMSGID.GtCltHandshake] = new MSG_Gt_Clt_Handshake();
			m_MessagePool[(int)CMSGID.CltGtHeartbeat] = new MSG_Clt_Gt_Heartbeat();
			m_MessagePool[(int)CMSGID.GtCltHeartbeat] = new MSG_Gt_Clt_Heartbeat();
			m_MessagePool[(int)CMSGID.GtCltDisconnect] = new MSG_Gt_Clt_Disconnect();
			m_MessagePool[(int)CMSGID.CltGLogin] = new MSG_Clt_G_Login();
			m_MessagePool[(int)CMSGID.GCltLoginResult] = new MSG_G_Clt_LoginResult();
			m_MessagePool[(int)CMSGID.GCltTestData] = new MSG_G_Clt_TestData();
			m_MessagePool[(int)SMSGID.CcCHandshake] = new MSG_Cc_C_Handshake();
			m_MessagePool[(int)SMSGID.CCcHandshake] = new MSG_C_Cc_Handshake();
			m_MessagePool[(int)SMSGID.CcCHeartbeat] = new MSG_Cc_C_Heartbeat();
			m_MessagePool[(int)SMSGID.CcCServerInfo] = new MSG_Cc_C_ServerInfo();
			m_MessagePool[(int)SMSGID.CCcServerInfo] = new MSG_C_Cc_ServerInfo();
			m_MessagePool[(int)SMSGID.CcCServerLost] = new MSG_Cc_C_ServerLost();
			m_MessagePool[(int)SMSGID.CCcServerLost] = new MSG_C_Cc_ServerLost();
			m_MessagePool[(int)SMSGID.CcCServerOk] = new MSG_Cc_C_ServerOk();
			m_MessagePool[(int)SMSGID.CCcServerOk] = new MSG_C_Cc_ServerOk();
			m_MessagePool[(int)SMSGID.CCcStop] = new MSG_C_Cc_Stop();
			m_MessagePool[(int)SMSGID.GtGtAHandshake] = new MSG_Gt_GtA_Handshake();
			m_MessagePool[(int)SMSGID.GtAGtHandshake] = new MSG_GtA_Gt_Handshake();
			m_MessagePool[(int)SMSGID.GtGtAHeartbeat] = new MSG_Gt_GtA_Heartbeat();
			m_MessagePool[(int)SMSGID.GtGtAClientClose] = new MSG_Gt_GtA_ClientClose();
			m_MessagePool[(int)SMSGID.GtAGtClientDisconnect] = new MSG_GtA_Gt_ClientDisconnect();
			m_MessagePool[(int)SMSGID.GtAGtClientMessage] = new MSG_GtA_Gt_ClientMessage();
			m_MessagePool[(int)SMSGID.GtAGtBroadcast] = new MSG_GtA_Gt_Broadcast();
			m_MessagePool[(int)SMSGID.GtAGtSetServerId] = new MSG_GtA_Gt_SetServerId();
			m_MessagePool[(int)SMSGID.DbcDbHandshake] = new MSG_Dbc_Db_Handshake();
			m_MessagePool[(int)SMSGID.DbDbcHandshake] = new MSG_Db_Dbc_Handshake();
			m_MessagePool[(int)SMSGID.DbcDbHeartbeat] = new MSG_Dbc_Db_Heartbeat();
			m_MessagePool[(int)SMSGID.DbcDbRequest] = new MSG_Dbc_Db_Request();
			m_MessagePool[(int)SMSGID.DbDbcResponse] = new MSG_Db_Dbc_Response();
			m_MessagePool[(int)SMSGID.HcHHandshake] = new MSG_Hc_H_Handshake();
			m_MessagePool[(int)SMSGID.HHcHandshake] = new MSG_H_Hc_Handshake();
			m_MessagePool[(int)SMSGID.HcHHeartbeat] = new MSG_Hc_H_Heartbeat();
			m_MessagePool[(int)SMSGID.GGHubTest] = new MSG_G_G_HubTest();
            //MESSAGE_END
            // 上面的注释定义必须保留，用来自动生成代码

            XSFCore.messageHelper = this;
        }

        public IMessage GetMessage(ushort nID)
        {
            return m_MessagePool[nID];
        }

        public static void CreateModule()
        {
            MessageModule module = new MessageModule();

            ModuleInit init = new ModuleInit();
            init.ID = 1;
            init.Name = "Message";
            init.NoWaitStart = true;

            XSFCore.Server.AddModule(module, init);
        }
    }
}

File: server-cpp/source/message/src/MessageModule.cpp
//////////////////////////////////////////////////////////////////////////
//
// 文件：message\src\MessageModule.cpp
// 作者：Xoen Xie
// 时间：2020/05/29
// 描述：消息模块
// 说明：
//
//////////////////////////////////////////////////////////////////////////

#include "MessageModule.h"
#include "DMessageInc.h"

namespace xsf_msg
{
    #define NEW_MESSAGE(_NAME, _ID)     m_Messages[_ID] = new MSG_##_NAME();

    bool MessageModule::Init(ModuleInit *pInit)
    {
        NEW_MESSAGE(Clt_Gt_Handshake, xsf_pbid::CMSGID::Clt_Gt_Handshake)
        NEW_MESSAGE(Clt_Gt_Handshake, xsf_pbid::CMSGID::Clt_Gt_Handshake)
        NEW_MESSAGE(Gt_Clt_Handshake, xsf_pbid::CMSGID::Gt_Clt_Handshake)
        NEW_MESSAGE(Clt_Gt_Heartbeat, xsf_pbid::CMSGID::Clt_Gt_Heartbeat)
        NEW_MESSAGE(Gt_Clt_Heartbeat, xsf_pbid::CMSGID::Gt_Clt_Heartbeat)
        NEW_MESSAGE(Gt_Clt_Disconnect, xsf_pbid::CMSGID::Gt_Clt_Disconnect)
        NEW_MESSAGE(Clt_G_Login, xsf_pbid::CMSGID::Clt_G_Login)
        NEW_MESSAGE(G_Clt_LoginResult, xsf_pbid::CMSGID::G_Clt_LoginResult)
        NEW_MESSAGE(G_Clt_TestData, xsf_pbid::CMSGID::G_Clt_TestData)
        NEW_MESSAGE(Cc_C_Handshake, xsf_pbid::SMSGID::Cc_C_Handshake)
        NEW_MESSAGE(C_Cc_Handshake, xsf_pbid::SMSGID::C_Cc_Handshake)
        NEW_MESSAGE(Cc_C_Heartbeat, xsf_pbid::SMSGID::Cc_C_Heartbeat)
        NEW_MESSAGE(Cc_C_ServerInfo, xsf_pbid::SMSGID::Cc_C_ServerInfo)
        NEW_MESSAGE(C_Cc_ServerInfo, xsf_pbid::SMSGID::C_Cc_ServerInfo)
        NEW_MESSAGE(Cc_C_ServerLost, xsf_pbid::SMSGID::Cc_C_ServerLost)
        NEW_MESSAGE(C_Cc_ServerLost, xsf_pbid::SMSGID::C_Cc_ServerLost)
        NEW_MESSAGE(Cc_C_ServerOk, xsf_pbid::SMSGID::Cc_C_ServerOk)
        NEW_MESSAGE(C_Cc_ServerOk, xsf_pbid::SMSGID::C_Cc_ServerOk)
        NEW_MESSAGE(C_Cc_Stop, xsf_pbid::SMSGID::C_Cc_Stop)
        NEW_MESSAGE(Gt_GtA_Handshake, xsf_pbid::SMSGID::Gt_GtA_Handshake)
        NEW_MESSAGE(GtA_Gt_Handshake, xsf_pbid::SMSGID::GtA_Gt_Handshake)
        NEW_MESSAGE(Gt_GtA_Heartbeat, xsf_pbid::SMSGID::Gt_GtA_Heartbeat)
        NEW_MESSAGE(Gt_GtA_ClientClose, xsf_pbid::SMSGID::Gt_GtA_ClientClose)
        NEW_MESSAGE(GtA_Gt_ClientDisconnect, xsf_pbid::SMSGID::GtA_Gt_ClientDisconnect)
        NEW_MESSAGE(GtA_Gt_ClientMessage, xsf_pbid::SMSGID::GtA_Gt_ClientMessage)
        NEW_MESSAGE(GtA_Gt_Broadcast, xsf_pbid::SMSGID::GtA_Gt_Broadcast)
        NEW_MESSAGE(GtA_Gt_SetServerId, xsf_pbid::SMSGID::GtA_Gt_SetServerId)

        XSFCore::mMessageHelper = this;
        
        return IModule::Init(pInit);
    }



    void MessageModule::Release(void)
    {
        for (int32 i = 0; i < xsf_pbid::SMSGID_Max; ++i)
        {
            if (m_Messages[i] != nullptr)
            {
                m_Messages[i]->Release();
                m_Messages[i] = nullptr;
            }
        }

        google::protobuf::ShutdownProtobufLibrary();

        delete this;
    }

    IMessage *MessageModule::GetMessage(uint16 nID)
    {
        if (nID >= xsf_pbid::SMSGID_Max)
        {
            XSF_ERROR("MessageModule::GetMessage nID:%u >= MSGID_Max", nID);
            return nullptr;
        }

        return m_Messages[nID];
    }

    void SetMessageModule(void)
    {
        MessageModule *pMessageModule = new MessageModule();

        ModuleInit *pInit = new ModuleInit();
        pInit->nModuleID = MESSAGE_MODULE_ID;
        strcpy(pInit->sName, "MessageModule");
        pInit->NoWaitStart = true;

        XSFCore::GetServer()->AddModule(pMessageModule, pInit);
    }
}
